### PR TITLE
HPCC-27963 Added roxiemem lock/unlock support

### PR DESCRIPTION
**Pull request by mckellyln**
Signed-off-by: Mark Kelly <mark.kelly@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by richardkchapman — APPROVED** [review]

File: roxie/ccd/ccdstate.cpp [hunk 1]
```diff
@@ -2588,12 +2588,30 @@ class CRoxiePackageSetManager : implements IRoxieQueryPackageManagerSet, impleme
...
             {
-                roxiemem::lockRoxieMem(true);
+                int srtn = roxiemem::lockRoxieMem(true);
+                if (!srtn)
+                {
+                    topology->setPropBool("@heapLockMemory", true);
+                    reply.append(" <heapLockMemory>1</heapLockMemory>\n");
+                }
+                else
+                    reply.appendf(" <Error=%d, unable to lock roxieMemory/>\n", srtn);
```
Comment (ghalliday) [inline roxie/ccd/ccdstate.cpp:2598]: This isn't valid xml - the receiver won't be able to parse it.

Maybe 
`<LockedMemory locked=%u [error=%s]/>
`?

**Review by ghalliday — CHANGES_REQUESTED** [review]

File: roxie/ccd/ccdstate.cpp [hunk 2]
```diff
@@ -2588,12 +2588,30 @@ class CRoxiePackageSetManager : implements IRoxieQueryPackageManagerSet, impleme
             }
             else if (strieq(queryName, "control:memlock"))
             {
-                roxiemem::lockRoxieMem(true);
+                int srtn = roxiemem::lockRoxieMem(true);
+                if (!srtn)
+                {
+                    topology->setPropBool("@heapLockMemory", true);
+                    reply.append(" <heapLockMemory>1</heapLockMemory>\n");
```
Comment (afishbeck) [inline roxie/ccd/ccdstate.cpp:2595]: I'd prefer we were consistent with the xml formatting.  In the successful replies the lock state is the content, but in the unsuccessful replies, the lock state is an attribute.

So for consistency, 

`reply.append(" <heapLockMemory locked='1'/>\n");`

would be better.

File: roxie/ccd/ccdstate.cpp [hunk 3]
```diff
@@ -2588,12 +2588,30 @@ class CRoxiePackageSetManager : implements IRoxieQueryPackageManagerSet, impleme
...
+                {
+                    topology->setPropBool("@heapLockMemory", false);
+                    reply.append(" <heapLockMemory>0</heapLockMemory>\n");
+                }
+                else
+                    reply.appendf(" <heapLockMemory locked='%d' error='unable to unlock roxieMemory:%d'/>\n", topology->getPropBool("@heapLockMemory"), srtn);
+            }
+            else if (strieq(queryName, "control:memquerylock"))
+            {
+                reply.appendf(" <heapLockMemory>%d</heapLockMemory>\n", topology->getPropBool("@heapLockMemory"));
```
Comment (afishbeck) [inline roxie/ccd/ccdstate.cpp:2614]: Based on above comments would make this:

reply.appendf(" <heapLockMemory locked='%d'/>\n", roxiemem::getRoxieMemLocked());

**Comment by mckellyln** [comment]
@afishbeck Thanks I've added code in a commit to address all your comments.

File: roxie/ccd/ccdstate.cpp [hunk 4]
```diff
@@ -2588,12 +2592,26 @@ class CRoxiePackageSetManager : implements IRoxieQueryPackageManagerSet, impleme
             }
             else if (strieq(queryName, "control:memlock"))
             {
-                roxiemem::lockRoxieMem(true);
+                int srtn = roxiemem::lockRoxieMem(true);
+                if (!srtn)
+                {
+                    topology->setPropBool("@heapLockMemory", true);
+                    reply.append(" <heapLockMemory locked='1'/>\n");
```
Comment (afishbeck) [inline roxie/ccd/ccdstate.cpp:2599]: Trivial, but typical naming convention has XML tag names starting with capitol, and attribute names starting with lower.

So `<HeapLockMemory locked='1'/>`

Comment (mckellyln) [reply roxie/ccd/ccdstate.cpp:2599]: ok.  I was trying to match the config name option case (heapLockMemory).  Not sure which is better ?

Comment (afishbeck) [reply roxie/ccd/ccdstate.cpp:2599]: In most places in our code I think xml tags are uppercase.  But in roxie control messages there are a few exceptions already, so it's up to you.

**Review by afishbeck — APPROVED** [review]

File: roxie/ccd/ccdstate.cpp (hunk 4 above)
Comment (mckellyln) [reply roxie/ccd/ccdstate.cpp:2599]: ok, I think I will leave it as is.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**